src/compiler/scala/tools/nsc/typechecker/Infer.scala

```diff
@@ -1434,28 +1435,42 @@ trait Infer extends Checkable {
     val Result  = mutable.LinkedHashMap
     type Result = mutable.LinkedHashMap[Symbol, Option[Type]]
 
-    def unapply(m: Result): Some[(List[Symbol], List[Type])] = Some(toLists(
-      (m collect {case (p, Some(a)) => (p, a)}).unzip  ))
+    def unapply(m: Result): Some[(List[Symbol], List[Type])] = {
+      val keys: ListBuffer[Symbol] = ListBuffer.empty
+      val vals: ListBuffer[Type]   = ListBuffer.empty
+      m foreach {
+        case (p, Some(a)) => keys += p ; vals += a
+        case _ =>
+      }
+      Some((keys.toList, vals.toList))
+    }
 
     object Undets {
-      def unapply(m: Result): Some[(List[Symbol], List[Type], List[Symbol])] = Some(toLists{
-        val (ok, nok) = m.map{case (p, a) => (p, a.getOrElse(null))}.partition(_._2 ne null)
-        val (okArgs, okTparams) = ok.unzip
-        (okArgs, okTparams, nok.keys)
-      })
+      def unapply(m: Result): Some[(List[Symbol], List[Type], List[Symbol])] = {
+        val okArgs: ListBuffer[Symbol] = ListBuffer.empty
+        val okTparams: ListBuffer[Type]   = ListBuffer.empty
+        val nokArs: ListBuffer[Symbol] = ListBuffer.empty
+        m foreach {
+          case (p, Some(a)) => okArgs += p ; okTparams += a
+          case (p, None)    => nokArs += p
+        }
+        Some((okArgs.toList, okTparams.toList, nokArs.toList))
+      }
     }
 
     object AllArgsAndUndets {
-      def unapply(m: Result): Some[(List[Symbol], List[Type], List[Type], List[Symbol])] = Some(toLists{
-        val (ok, nok) = m.map{case (p, a) => (p, a.getOrElse(null))}.partition(_._2 ne null)
-        val (okArgs, okTparams) = ok.unzip
-        (okArgs, okTparams, m.values.map(_.getOrElse(NothingTpe)), nok.keys)
-      })
+      def unapply(m: Result): Some[(List[Symbol], List[Type], List[Type], List[Symbol])] = {
+        val oksKeys: ListBuffer[Symbol] = ListBuffer.empty
+        val nokKeys: ListBuffer[Symbol] = ListBuffer.empty
+        val okTparams: ListBuffer[Type] = ListBuffer.empty
+        val valsOrNothing: ListBuffer[Type] = ListBuffer.empty
+        m foreach {
+          case (p, Some(a) ) => oksKeys += p ; valsOrNothing += a         ; okTparams += a
```

Comment:
Having in mind the comment above, I hope that using a matrix-like alignment will make the code more readable.